### PR TITLE
Added copy ctor to worlddata and robot

### PR DESCRIPTION
**Pull request by Yuhanun**
Both are now copyable

Suggestion for the tree:

Copy the worlddata once, set it in the origin, then pass worlddataviews to  that

### CODE REVIEW

File: src/world_new/WorldData.cpp [hunk 1]
```diff
@@ -9,58 +9,79 @@
...
+        return us;
     }
-    for (auto &each : others) {
-        auto _ptr = &robots.emplace_back(feedback, each, Team::them, getBall());
-        them.emplace_back(_ptr);
-        robotsNonOwning.emplace_back(_ptr);
+
+    std::vector<view::RobotView> const &WorldData::getThem() const noexcept { return them; }
+
+    std::vector<rtt::world_new::robot::Robot> const &WorldData::getRobots() const noexcept { return robots; }
```
Comment (emielsteerneman) [inline src/world_new/WorldData.cpp:46]: Is it safe to provide a function that gives access to mutable objects?

**Review by emielsteerneman — APPROVED** [review]

File: src/world_new/WorldData.cpp (hunk 1 above)
Comment (Yuhanun) [reply src/world_new/WorldData.cpp:46]: Resolved.

File: include/roboteam_ai/world_new/WorldData.hpp [hunk 2]
```diff
@@ -97,6 +98,11 @@ class WorldData {
      */
     WorldData &operator=(WorldData &&) = default;
 
+    /**
+     * Does exactly what it says, it sets the internal vectors for the views
+     */
+    void setViewVectors() noexcept;
```
Comment (mrlukasbos) [inline include/roboteam_ai/world_new/WorldData.hpp:104]: shouldn't this function be private?

File: src/world_new/WorldData.cpp [hunk 3]
```diff
@@ -9,58 +9,79 @@
...
-uint64_t WorldData::getTime() const noexcept { return time; }
+    WorldData::WorldData(WorldData const & other) noexcept :
+            robots{ other.robots }, ball{ other.ball }{
+        setViewVectors();
+    }
 
-std::vector<view::RobotView> const &WorldData::getRobotsNonOwning() const noexcept { return robotsNonOwning; }
+    void WorldData::setViewVectors() noexcept {
+        /*
+         * Reserve the us and them vectors, making sure they are not being resized (which will invalidate the references)
```
Comment (mrlukasbos) [inline src/world_new/WorldData.cpp:88]: where does this reserve happen?

**Review by mrlukasbos — CHANGES_REQUESTED** [review]
It looks like you don't reserve us and them anymore

File: src/world_new/WorldData.cpp (hunk 3 above)
Comment (Yuhanun) [reply src/world_new/WorldData.cpp:88]: Oh yeah, nice catch, ill take a look at that tomorrow

**Comment by Yuhanun** [comment]
Resolved the issue where iterators would be invalidated.

File: include/roboteam_ai/world_new/WorldData.hpp (hunk 2 above)
Comment (Yuhanun) [reply include/roboteam_ai/world_new/WorldData.hpp:104]: Doesn't really matter, it doesn't fuck with the actual robots so yeah, but I'll make it private.

File: src/world_new/WorldData.cpp [hunk 4]
```diff
@@ -9,58 +9,83 @@
...
+        robots.reserve(amountUs + amountThem);
+
+        for (auto &each : ours) {
+            robots.emplace_back(feedback, each, Team::us, getBall());
+        }
+        for (auto &each : others) {
+            robots.emplace_back(feedback, each, Team::them, getBall());
+        }
+
+
```
Comment (mrlukasbos) [inline src/world_new/WorldData.cpp]:
```suggestion
us.reserve(amountUs);
them.reserve(amountThem);
```

**Review by mrlukasbos — CHANGES_REQUESTED** [review]
Im not sure but you still may have UB in the us and them vectors. When constructing a worldData from a protoMessage. The setViewVector then fills us and them with robots from the 'robots' vector.

**Comment by Yuhanun** [comment]
is that what you meant?

**Pull request merged**